app/controllers/bz_controller.rb

```diff
@@ -174,17 +174,77 @@ def user_retained_data
   def set_user_retained_data
     result = RetainedData.where(:user_id => @current_user.id, :name => params[:name])
     data = nil
+    was_new = false
     if result.empty?
       data = RetainedData.new()
       data.user_id = @current_user.id
       data.path = request.referrer
       data.name = params[:name]
+      was_new = true
     else
       data = result.first
     end
 
     data.value = params[:value]
     data.save
+
+    # now that the user's work is safely saved, we will go back and do addon work
+    # like micrograding
+
+    if was_new
+      course_id = request.referrer[/\/courses\/(\d+)\//, 1]
+      if course_id
+
+        course = Course.find(course_id)
+
+        # counting the total number of magic fields is a really slow operation since it needs to
+        # scan the whole course. Thus, it is aggressively cached. Major changes to a course in
+        # the middle of a term can thus throw off scores (since the existing points don't adapt to
```

Comment:
I think this is fine.  We should let the content designers know about this caveat though.